diff --git a/proton b/proton
index 120a208..a7b6842 100755
--- a/proton
+++ b/proton
@@ -1701,6 +1701,7 @@ class Session:
         self.check_environment("PROTON_NO_D3D10", "nod3d10")
         self.check_environment("PROTON_NO_ESYNC", "noesync")
         self.check_environment("PROTON_NO_FSYNC", "nofsync")
+        self.check_environment("PROTON_NO_NTSYNC", "nontsync")
         self.check_environment("PROTON_FORCE_LARGE_ADDRESS_AWARE", "forcelgadd")
         self.check_environment("PROTON_OLD_GL_STRING", "oldglstr")
         self.check_environment("PROTON_ENABLE_WAYLAND", "enablewayland")
@@ -1740,6 +1741,9 @@ class Session:
         else:
             self.env["WINEFSYNC"] = "1"
 
+        if "nontsync" in self.compat_config:
+            self.env["WINE_DISABLE_FAST_SYNC"] = "1"
+
         if "oldglstr" in self.compat_config:
             #mesa override
             self.env["MESA_EXTENSION_MAX_YEAR"] = "2003"
