diff --git a/meson.build b/meson.build
index 64b6803..55b38d9 100644
--- a/meson.build
+++ b/meson.build
@@ -10,10 +10,8 @@ cc = meson.get_compiler('c')
 dxvk_is_msvc = cpp.get_argument_syntax() == 'msvc'
 
 compiler_args = [
-  '-msse',
-  '-msse2',
-  '-msse3',
-  '-mfpmath=sse',
+  '-O3',
+  '-march=native',
   '-Wimplicit-fallthrough',
   # gcc
   '-Wno-missing-field-initializers',
@@ -93,7 +91,7 @@ if platform == 'windows'
     endif
   else
     # setup file alignment + enable PDB output for MSVC builds
-    # PDBs are useful for Windows consumers of DXVK 
+    # PDBs are useful for Windows consumers of DXVK
     compiler_args += [
       '/Z7'
     ]
@@ -162,7 +160,7 @@ else
   if (not lib_sdl3.found() and not lib_sdl2.found() and not lib_glfw.found())
     error('SDL3, SDL2, or GLFW are required to build dxvk-native')
   endif
-  
+
   dxvk_name_prefix = 'dxvk_'
   dxvk_pkg_prefix = 'dxvk-'
 
