diff --git a/Makefile.am b/Makefile.am
index dccd89dd..6396f052 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -652,7 +652,7 @@ dummy-vkd3d-version:
 ## Cross-compile tests
 cross_implibs = crosslibs/d3d12
 CROSS_CPPFLAGS = -I$(srcdir)/include -I$(srcdir)/include/private -I$(builddir)/include -I$(builddir)/tests
-CROSS_CFLAGS = -g -O2 -Wall -municode ${CROSS_CPPFLAGS} -Wa,-I$(srcdir)/demos \
+CROSS_CFLAGS = -g -O3 -mtune=znver2 -Wall -municode ${CROSS_CPPFLAGS} -Wa,-I$(srcdir)/demos \
 	-D_WIN32_WINNT=0x0600 -D__USE_MINGW_ANSI_STDIO=0 -DVKD3D_CROSSTEST=1
 EXTRA_DIST += $(cross_implibs:=.cross32.def) $(cross_implibs:=.cross64.def)
 EXTRA_DIST += tests/driver.c tests/shader_runner_d3d11.c tests/shader_runner_d3d9.c
diff --git a/configure.ac b/configure.ac
index 884d9238..72faca62 100644
--- a/configure.ac
+++ b/configure.ac
@@ -62,7 +62,7 @@ gl_LD_VERSION_SCRIPT
 dnl Check compiler specific flags
 AC_SUBST([VKD3D_CFLAGS])
 AS_IF([test "x${GCC}" = "xyes"],
-      [VKD3D_CFLAGS="-Wall -pipe -D_WIN32_WINNT=0x0600"
+      [VKD3D_CFLAGS="-O3 -mtune=znver2 -Wall -pipe -D_WIN32_WINNT=0x0600"
       VKD3D_CHECK_CFLAGS([-std=c99])
       VKD3D_CHECK_CFLAGS([-flto=auto])
       VKD3D_CHECK_CFLAGS([-fvisibility=hidden])
