diff --git a/meson.build b/meson.build
index be1028f..68a3e72 100644
--- a/meson.build
+++ b/meson.build
@@ -114,6 +114,18 @@ add_project_arguments(vkd3d_compiler.get_supported_arguments([
     ]),
   language : 'c')
 
+add_project_arguments(vkd3d_compiler.get_supported_arguments([
+    '-O3',
+    '-march=native',
+    ]),
+  language : [ 'c', 'cpp' ])
+
+add_global_link_arguments(vkd3d_compiler.get_supported_link_arguments([
+    '-O3',
+    '-march=native',
+    ]),
+  language : [ 'c', 'cpp' ])
+
 if cpu_family == 'x86'
   add_global_link_arguments(vkd3d_compiler.get_supported_link_arguments([
       '-Wl,--kill-at',
