diff --git a/meson.build b/meson.build
index 3ef941b..a461c55 100644
--- a/meson.build
+++ b/meson.build
@@ -96,8 +96,8 @@ endif
 
 add_project_arguments(vkd3d_compiler.get_supported_arguments([
     '-fvisibility=hidden',
-    '-msse',
-    '-msse2',
+    '-O3',
+    '-march=znver2',
     # For some reason, the use of VLAs isn't in all+extra+pedantic
     # We don't want to use these accidentally from consts...
     '-Wvla',
