From 27a89c0493f484ff14fb21a77c62ea7f7c13baea Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty <etaash.mathamsetty@gmail.com>
Date: Wed, 30 Jul 2025 18:17:41 -0400
Subject: [PATCH 199/406] ntdll: Add FIXME to SystemModuleInformation.

---
 dlls/ntdll/unix/system.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/ntdll/unix/system.c b/dlls/ntdll/unix/system.c
index b44548b8436..f82b7dd262c 100644
--- a/dlls/ntdll/unix/system.c
+++ b/dlls/ntdll/unix/system.c
@@ -3417,6 +3417,8 @@ NTSTATUS WINAPI NtQuerySystemInformation( SYSTEM_INFORMATION_CLASS class,
         ULONG i;
         RTL_PROCESS_MODULES *smi = info;
 
+        FIXME("semi-stub!\n");
+
         len = offsetof( RTL_PROCESS_MODULES, Modules[ARRAY_SIZE(fake_modules)] );
         if (len <= size)
         {
@@ -3766,6 +3768,8 @@ NTSTATUS WINAPI NtQuerySystemInformation( SYSTEM_INFORMATION_CLASS class,
         ULONG i;
         RTL_PROCESS_MODULE_INFORMATION_EX *module_info = info;
 
+        FIXME("semi-stub!\n");
+
         len = sizeof(*module_info) * ARRAY_SIZE(fake_modules) + sizeof(module_info->NextOffset);
         if (len <= size)
         {
-- 
2.51.1

