From bf0096f390d1413bca4f42e02b59c063592c0333 Mon Sep 17 00:00:00 2001
From: Alexandre Julliard <julliard@winehq.org>
Date: Mon, 31 Mar 2025 12:19:16 +0200
Subject: [PATCH 250/339] comdlg32: Delay import winspool

To avoid creating an extra known dll section.
---
 dlls/comdlg32/Makefile.in | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/comdlg32/Makefile.in b/dlls/comdlg32/Makefile.in
index 7cf3b714b2b..c298c0801cc 100644
--- a/dlls/comdlg32/Makefile.in
+++ b/dlls/comdlg32/Makefile.in
@@ -1,8 +1,8 @@
 EXTRADEFS = -D_COMDLG32_
 MODULE    = comdlg32.dll
 IMPORTLIB = comdlg32
-IMPORTS   = uuid shell32 shlwapi comctl32 winspool user32 gdi32 advapi32
-DELAYIMPORTS = ole32
+IMPORTS   = uuid shell32 shlwapi comctl32 user32 gdi32 advapi32
+DELAYIMPORTS = ole32 winspool
 
 SOURCES = \
 	cdlg32.c \
-- 
2.50.1

