From 7dc387796bbd91a853c8d82c7758f231be9ba5bb Mon Sep 17 00:00:00 2001
From: Jacek Caban <jacek@codeweavers.com>
Date: Sat, 8 Feb 2025 20:54:20 +0100
Subject: [PATCH 273/365] ntdll: Pass importer pointer to build_import_name.

---
 dlls/ntdll/loader.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/dlls/ntdll/loader.c b/dlls/ntdll/loader.c
index 3683093119f..cb11676bd7d 100644
--- a/dlls/ntdll/loader.c
+++ b/dlls/ntdll/loader.c
@@ -769,11 +769,11 @@ static NTSTATUS get_apiset_target( const API_SET_NAMESPACE *map, const API_SET_N
 /**********************************************************************
  *	    build_import_name
  */
-static NTSTATUS build_import_name( WCHAR buffer[256], const char *import, int len )
+static NTSTATUS build_import_name( WINE_MODREF *importer, WCHAR buffer[256], const char *import, int len )
 {
     const API_SET_NAMESPACE *map = NtCurrentTeb()->Peb->ApiSetMap;
     const API_SET_NAMESPACE_ENTRY *entry;
-    const WCHAR *host = current_modref ? current_modref->ldr.BaseDllName.Buffer : NULL;
+    const WCHAR *host = importer ? importer->ldr.BaseDllName.Buffer : NULL;
     UNICODE_STRING str;
 
     while (len && import[len-1] == ' ') len--;  /* remove trailing spaces */
@@ -948,7 +948,7 @@ static FARPROC find_forwarded_export( HMODULE module, const char *forward, LPCWS
     FARPROC proc = NULL;
 
     if (!end) return NULL;
-    if (build_import_name( mod_name, forward, end - forward )) return NULL;
+    if (build_import_name( importer, mod_name, forward, end - forward )) return NULL;
 
     if (!(wm = find_basename_module( mod_name )))
     {
@@ -1162,7 +1162,8 @@ static BOOL import_dll( WINE_MODREF *wm, const IMAGE_IMPORT_DESCRIPTOR *descr, L
         name = "ntdll.dll";
     }
 
-    status = build_import_name( buffer, name, len );
+    status = build_import_name( wm, buffer, name, len );
+
     if (!status) status = load_dll( load_path, buffer, 0, &wmImp, system );
 
     if (status)
-- 
2.50.1

