From f9915154460da769ee6189bd9bf4a320886c9909 Mon Sep 17 00:00:00 2001
From: Jacek Caban <jacek@codeweavers.com>
Date: Sat, 8 Feb 2025 20:55:37 +0100
Subject: [PATCH 274/365] ntdll: Use NULL importer in LdrGetProcedureAddress.

---
 dlls/ntdll/loader.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/ntdll/loader.c b/dlls/ntdll/loader.c
index cb11676bd7d..e7bbd7baeda 100644
--- a/dlls/ntdll/loader.c
+++ b/dlls/ntdll/loader.c
@@ -2124,8 +2124,8 @@ NTSTATUS WINAPI LdrGetProcedureAddress(HMODULE module, const ANSI_STRING *name,
     else if ((exports = RtlImageDirectoryEntryToData( module, TRUE,
                                                       IMAGE_DIRECTORY_ENTRY_EXPORT, &exp_size )))
     {
-        void *proc = name ? find_named_export( module, exports, exp_size, name->Buffer, -1, NULL, current_modref )
-                          : find_ordinal_export( module, exports, exp_size, ord - exports->Base, NULL, current_modref );
+        void *proc = name ? find_named_export( module, exports, exp_size, name->Buffer, -1, NULL, NULL )
+                          : find_ordinal_export( module, exports, exp_size, ord - exports->Base, NULL, NULL );
         if (proc && !is_hidden_export( proc ))
         {
             *address = proc;
-- 
2.50.1

