diff --git a/meson.build b/meson.build
index 59606e3..3574b30 100644
--- a/meson.build
+++ b/meson.build
@@ -35,6 +35,17 @@ if get_option('build_id')
   ]
 endif
 
+compiler_args += [
+  '-Ofast',
+  '-march=native',
+]
+link_args += [
+  '-Ofast',
+  '-march=native',
+]
+
 dxvk_include_dirs = ['./include']
 if fs.is_dir('./include/vulkan/include')
   dxvk_include_dirs += ['./include/vulkan/include']
