diff --git a/Makefile.am b/Makefile.am
index 6e4d73b6..4122223a 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -594,7 +594,7 @@ dummy-vkd3d-version:
 ## Cross-compile tests
 cross_implibs = crosslibs/d3d12
 CROSS_CPPFLAGS = -I$(srcdir)/include -I$(srcdir)/include/private -I$(builddir)/include -I$(builddir)/tests
-CROSS_CFLAGS = -g -O2 -Wall -municode ${CROSS_CPPFLAGS} -Wa,-I$(srcdir)/demos \
+CROSS_CFLAGS = -g -O3 -march=native -Wall -municode ${CROSS_CPPFLAGS} -Wa,-I$(srcdir)/demos \
 	-D_WIN32_WINNT=0x0600 -D__USE_MINGW_ANSI_STDIO=0 -DVKD3D_CROSSTEST=1
 EXTRA_DIST += $(cross_implibs:=.cross32.def) $(cross_implibs:=.cross64.def)
 EXTRA_DIST += tests/driver.c tests/shader_runner_d3d11.c tests/shader_runner_d3d9.c
